tokio-executor/src/task/harness.rs

```diff
@@ -0,0 +1,592 @@
+            }
+        }
+    }
+
+    /// Return `true` if the task structure should be deallocated
+    fn transition_to_complete(&mut self, join_interest: bool) -> Snapshot {
+        let res = self.header().state.transition_to_complete();
+
+        self.notify_join_handle(join_interest, res);
+
+        // Transition to complete last to ensure freeing does
+        // not happen until the above work is done.
+        res
+    }
+
+    /// Return `true` if the task structure should be deallocated
+    fn transition_to_released(&mut self, join_interest: bool) -> Snapshot {
+        if join_interest {
+            let res1 = self.transition_to_complete(join_interest);
+
+            // At this point, the join waker may not be changed. Once we perform
+            // `release_task` we may no longer read from the struct but we
+            // **may** be responsible for dropping the waker. We do an
+            // optimistic read here.
+            let (join_waker, check) = unsafe { self.read_join_waker() };
+
+            let res2 = self.header().state.release_task();
+
+            if res1.has_join_waker() && !res2.is_join_interested() {
+                debug_assert!(res2.has_join_waker());
+
+                // Its our responsibility to drop the waker
+                check.check();
+                unsafe {
+                    let _ = join_waker.assume_init();
+                }
+            }
+
+            res2
+        } else {
+            self.header().state.transition_to_released()
+        }
+    }
+
+    fn notify_join_handle(&mut self, join_interest: bool, res: Snapshot) {
+        if join_interest {
+            if !res.is_join_interested() {
+                debug_assert!(!res.has_join_waker());
+
+                // The join handle dropped interest before we could release
+                // the output. We are now responsible for releasing the
+                // output.
+                unsafe {
+                    self.drop_out();
+                }
+            } else if res.has_join_waker() {
+                if res.is_canceled() {
+                    // The join handle will set the output to Cancelled without
+                    // attempting to read the output. We must drop it here.
+                    unsafe {
+                        self.drop_out();
+                    }
+                }
+
+                // Notify the join handle. The previous transition obtains the
+                // lock on the waker cell.
+                unsafe {
+                    self.wake_join();
+                }
+            }
+        }
+    }
+
+    fn might_drop_join_waker_on_release(&self) -> bool {
+        unsafe {
+            let next = *self.header().queue_next.get() as usize;
+            next & 1 == 1
+        }
+    }
+
+    fn set_might_drop_join_waker_on_release(&self) {
+        unsafe {
+            // The next field must be null at this point
+            debug_assert!((*self.header().queue_next.get()).is_null());
```

Comment:
Nit/TIOLI: might be good if some of these debug assertions that assert fairly complex expressions had messages as well?